### PR TITLE
Opening a file in Read mode now returns a read-only stream

### LINKED ISSUE #230: MockFileInfo.OpenWrite returns stream with read access.
This is not the behaviour I receive when opening a filestream. My unit test passed but the code failed at runtime. I expected `OpenWrite()` to return a stream with no read access, and to throw the appropriate exception when a read is attempted.

image

### PR DESCRIPTION
**Pull request by Shtong**
Fixes #230

### CODE REVIEW

**Comment by fgreinacher** [comment]
@Shtong I know this PR is pretty old, but in case you still want to work in this, please update your branch and address the review comments. If you don’t want to work on this someone else might be able to help get this in :)

**Comment by Shtong** [comment]
I'll try to see if I can back into that during the week-end

File: TestingHelpers/MockFile.cs [hunk 1]
```diff
@@ -385,12 +385,14 @@ public override Stream Open(string path, FileMode mode, FileAccess access, FileS
...
-            var stream = OpenWrite(path);
-
-            if (mode == FileMode.Append)
-                stream.Seek(length, SeekOrigin.Begin);
+            
+            MockFileStream.StreamType streamType = MockFileStream.StreamType.WRITE;
+            if (access == FileAccess.Read)
+                streamType = MockFileStream.StreamType.READ;
+            else if (mode == FileMode.Append)
+                streamType = MockFileStream.StreamType.APPEND;
```
[the remark this answers is not in the document]
Comment (Shtong) [reply TestingHelpers/MockFile.cs]: Seeking is already handled by the `MockFileStream` constructor called later

**Comment by Shtong** [comment]
All done.

Note: the changes I made revealed a bug in the `MockFile_OpenWrite_ShouldCreateNewFiles` unit test. Because the folder in which the new file was requested didn't exist, `OpenWrite` should have thrown an exception, which is not the case in the master branch. The changes I made fixed this bug, so I fixed the unit tests to check for both scenarios (working with existing folder, and throwing with missing folder)

File: TestingHelpers/MockFileStream.cs [hunk 2]
```diff
@@ -38,6 +38,15 @@ public MockFileStream(IMockFileDataAccessor mockFileDataAccessor, string path, S
                 }
                 mockFileDataAccessor.AddFile(path, new MockFileData(new byte[] { }));
             }
+
+            canWrite = streamType != StreamType.READ;
+        }
+
+        private bool canWrite = true;
```
Comment (fgreinacher) [inline TestingHelpers/MockFileStream.cs]: Please move this field to the top, where the other fields are.

File: TestingHelpers/MockFileStream.cs [hunk 3]
```diff
@@ -38,6 +38,15 @@ public MockFileStream(IMockFileDataAccessor mockFileDataAccessor, string path, S
...
+
+        public override bool CanWrite
```
Comment (fgreinacher) [inline TestingHelpers/MockFileStream.cs]: `public override bool CanWrite => canWrite` ?

File: TestHelpers.Tests/MockFileTests.cs [hunk 4]
```diff
@@ -667,6 +678,26 @@ public void MockFile_Replace_ShouldThrowIfDestinationFileDoesNotExist()
 
             Assert.Throws<FileNotFoundException>(() => fileSystem.File.Replace(path1, path2, null));
         }
+
+        [Test]
+        public void MockFile_OpenRead_ShouldNotReturnWritableStream()
```
Comment (fgreinacher) [inline TestHelpers.Tests/MockFileTests.cs]: `...ShouldReturnReadOnlyStream`?

File: TestHelpers.Tests/MockFileTests.cs [hunk 5]
```diff
@@ -423,6 +424,16 @@ public void MockFile_ReadAllText_ShouldReturnTheOriginalContentWhenTheFileContai
...
 
+        [Test]
+        public void MockFile_OpenWrite_ShouldNotCreateFolders()
+        {
+            string filePath = XFS.Path(@"c:\something\demo.txt"); // c:\something does not exist: OpenWrite should fail
+            var fileSystem = new MockFileSystem();
+
+            Assert.Throws<DirectoryNotFoundException>(() => fileSystem.File.OpenWrite(filePath));
+        }
+
```
Comment (fgreinacher) [inline TestHelpers.Tests/MockFileTests.cs]: Remove superfluous empty line

**Comment by fgreinacher** [comment]
Great - thanks!

**Pull request merged**